src/allmydata/testing/web.py

```diff
@@ -0,0 +1,225 @@
+import hashlib
+
+import attr
+
+from hyperlink import DecodedURL
+
+from twisted.web.resource import (
+    Resource,
+)
+from twisted.web.iweb import (
+    IBodyProducer,
+)
+from twisted.internet.defer import (
+    succeed,
+)
+
+from treq.client import (
+    HTTPClient,
+    FileBodyProducer,
+)
+from treq.testing import (
+    RequestTraversalAgent,
+)
+from zope.interface import implementer
+
+import allmydata.uri
+from allmydata.util import (
+    base32,
+)
+
+
+class _FakeTahoeRoot(Resource, object):
+    """
+    This is a sketch of how an in-memory 'fake' of a Tahoe
+    WebUI. Ultimately, this will live in Tahoe
+    """
+
+    def __init__(self, uri=None):
+        Resource.__init__(self)  # this is an old-style class :(
+        self._uri = uri
+        self.putChild(b"uri", self._uri)
+
+    def add_data(self, kind, data):
+        return self._uri.add_data(kind, data)
+
+
+@attr.s
+class _FakeCapability(object):
+    """
+    """
+    data=attr.ib()
+
+
+KNOWN_CAPABILITIES = [
+    getattr(allmydata.uri, t).BASE_STRING
+    for t in dir(allmydata.uri)
+    if hasattr(getattr(allmydata.uri, t), 'BASE_STRING')
+]
+
+
+def capability_generator(kind):
+    """
+    :param str kind: the kind of capability, like `URI:CHK`
+
+    :returns: a generator that yields new capablities of a particular
+        kind.
+    """
+    if kind not in KNOWN_CAPABILITIES:
+        raise ValueError(
+            "Unknown capability kind '{} (valid are {})'".format(
+                kind,
+                ", ".join(KNOWN_CAPABILITIES),
+            )
+        )
+    # what we do here is to start with empty hashers for the key and
+    # ueb_hash and repeatedly feed() them a zero byte on each
+    # iteration .. so the same sequence of capabilities will always be
+    # produced. We could add a seed= argument if we wanted to produce
+    # different sequences.
+    number = 0
+    key_hasher = hashlib.new("sha256")
+    ueb_hasher = hashlib.new("sha256")
+
+    # capabilities are "prefix:<128-bits-base32>:<256-bits-base32>:N:K:size"
+    while True:
+        number += 1
+        key_hasher.update("\x00")
+        ueb_hasher.update("\x00")
+
+        key = base32.b2a(key_hasher.digest()[:16])  # key is 16 bytes
+        ueb_hash = base32.b2a(ueb_hasher.digest())  # ueb hash is 32 bytes
+
+        cap = u"{kind}:{key}:{ueb_hash}:{n}:{k}:{size}".format(
+            kind=kind,
+            key=key,
+            ueb_hash=ueb_hash,
+            n=1,
+            k=1,
+            size=number * 1000,
+        )
+        yield cap.encode("ascii")
+
+
+class _FakeTahoeUriHandler(Resource, object):
+    """
+    """
+
+    isLeaf = True
+    _data = None
```

Comment:
If `_data` were initialized to an empty dict a bunch of `... is None` checks would disappear from the implementation.  Since `object` is being mixed in here I think attrs might be usable, then `attr.ib(default=Factory(dict))`.

Reply:
Does that have any implications for `Resource` etc and twisted-web?

Reply:
(Seems to be fine ...)